[WIP] libssh backend

tiennou opened the pull request:
This adds support for libssh as a possible backend for the SSH transport. The code is untested — the test suite runs fine locally, but I want to see how CI reacts to it as I don't expect everything SSH to be covered by it (`keyboard-interactive` 🙄).

tiennou commented on src/transports/ssh_backend.h:
```diff
@@ -0,0 +1,477 @@
+					const char *instr = ssh_userauth_kbdint_getinstruction(session);
+					int num_prompts = ssh_userauth_kbdint_getnprompts(session);
+					size_t i;
+					ssh_userauth_kbdint_getprompt(session, i, NULL);
```
This is the API endpoint I'm having issues with — the public callback with the libssh2 types — with the libssh API.

 I'm not sure how to best proceed forward from here :
- break the API and make a more generic callback, with no exposed types from the backends (the new `git_ssh_` types).
- provide a libssh-specific version of it, passing only the session. This is contrary to the guideline that the API should not depend on preprocessing, and would cause Docurium to miss it (as an example).
- make the SSH session accessible, so we don't have to wrap so closely libssh2, and deprecate anything that's not really needed.

Any suggestions ?

tiennou commented on src/transports/ssh_backend.h:
```diff
@@ -0,0 +1,477 @@
+				assert(c->username);
+				assert(c->privatekey);
+
+				rc = libssh2_userauth_publickey_frommemory(
```
This is TBD.

tiennou commented:
I forgot to tag the crux of the issue in the PR's cover, but please see #5225 for context.

To clarify my perspective on #5307, ideally `git_cert` could have a `git_cert_ssh_cred` opaque object, or something like it, which would allow the user's code to just use the capabilities of their chosen backend, to access key material, auth-methods, or SSH channels directly.

This is not currently handled here, because I hit one case where the interface we already provide cannot be ported to libssh (the custom signing mechanism), so here's a summary of "currently known issues caused by libssh2": apparently it's gone from Debian (#5225), some requested features are missing (#5258, #5219, #4452), or depend on the libssh2's backend and key types supported (#4289, OpenSSL/libgcrypt only), and IIRC there's another issue…

cuviper commented on CMakeLists.txt:
```diff
@@ -52,7 +52,7 @@ OPTION(TAGS				"Generate tags"						OFF)
 OPTION(ENABLE_TRACE			"Enables tracing support"				OFF)
 OPTION(LIBGIT2_FILENAME			"Name of the produced binary"				OFF)
-OPTION(USE_SSH				"Link with libssh2 to enable SSH support"		 ON)
+OPTION(USE_SSH				"Enable SSH support. Can be set to a specific backend (libssh2)" ON)
```
I think this needs to be `SET` to properly cache the backend value, like: 

```suggestion
   SET(USE_SSH				"ON" CACHE STRING "Enable SSH support. Can be set to a specific backend (libssh2)")
```

cuviper commented on src/transports/ssh.h:
```diff
@@ -8,7 +8,73 @@
+	LIBSSH2_CHANNEL *channel;
+};
+
+#define git_ssh_channel_close(c) libssh2_channel_close(c->channel)
```
```suggestion
#define git_ssh_channel_close(c) libssh2_channel_close(c)
```

cuviper commented on src/transports/ssh.h:
```diff
@@ -8,7 +8,73 @@
+};
+
+#define git_ssh_channel_close(c) libssh2_channel_close(c->channel)
+#define git_ssh_channel_free(c) libssh2_channel_free(c->channel)
```
```suggestion
#define git_ssh_channel_free(c) libssh2_channel_free(c)
```

cuviper commented on src/transports/ssh_libssh2.c:
```diff
@@ -0,0 +1,376 @@
+				rc = libssh2_userauth_keyboard_interactive(
+					s->session, c->username, git_ssh__kbdint_cb);
+
+				*abstract = old_abstract;
```
I see you added `old_abstract` in commit 98c9f6a6, but its declaration got lost in the following commit 749f3440.

extrawurst commented:
@tiennou is this still worked on?

tiennou commented:
@extrawurst I've rebased it, updated CI, and the only showstopper now is that the `libssh` backend seem to require some system-level configuration files to exist.

@cuviper Thanks for the review, much appreciated ❤️. Issues (apart from the runtime config ones, and unforeseen bugs 😉) should be fixed.

carlosmn commented:
> This is the API endpoint I'm having issues with — the public callback with the libssh2 types — with the libssh API.

There's a fundamental conflict between the goals of the library and the realities of how different all these libraries and network protocols are as we both want to

1. have the same interface for performing Git operations over an arbitrary network protocol; and
2. allow the caller to react in whichever is the best way in their environment (language, user interaction)

The former goal means things like auth can get awkward as most HTTP auth doesn't look like SSH auth (and some HTTP auth doesn't look like a lot of other HTTP auth) but the latter means that we want to avoid processing the data too much when the end-user application (or whatever ends up calling the library) needs to interact with the library implementing the network protocol. This is made even more important by the fact that we want to allow callers of the library to plug in their own network layer and then they can devise an arbitrarily complex authentication method as long as both agree.

The wish to have the caller in control but also make it easy to perform the most common operations leads to weird things in the API like libssh2- and curl-specific which try to give the caller all the info they might need short of handing over a pointer to whatever context/handle each library has. At the same time it's rather disappointing how the information is presented by some libraries (it seems libcurl does the data processing we're trying not to do here), but that's a different discussion.

As far as your question about the callback types having specific libssh2 structures, I remember doing this due to the methods needed for the more complex interactions like keyboard-interactive, seem to be tied enough to libssh2 that did not seem realistic that we could provide some generic method for this. And from what I can tell, libssh has a rather different mechanism for doing this. Maybe it's not very sensible for this library to support this.

The same goes for certs where we only got two hashes from libssh2. It looks like the issue described in #5258 and #4452  has been fixed now, though it's not super helpful now that libssh2 isn't in some distros, I suppose. But we do have a place to put it so the caller can decide for themselves how they want to hash the cert if they need to. libssh expects you to ask for which hash you're interested in, which makes sense for it, but I don't know how much it makes for us to expose that detail.

For what it's worth, I have come to think that the design with the callbacks for host and user authentication were a mistake, but changing it to work with returning control to the user by returning is not a small task.

Ultimately, some of this probably isn't solvable as part of this library. We can try to keep it working with the features we have (and maybe not bother with keyboard-interactive for libssh?) and with libssh we can even fill in the info as to whether "we" think the ssh server is something you should trust. But we're still just not going to provide functionality for managing your `known_hosts` file. If an app does want to provide that, they're going to have to get dirty with libssh themselves or perhaps provide their own transport that talks to their app via some other means.

All this to say that attaching auth logic into libgit2 is always going to leave us with something somewhat annoying and the real solution is probably to provide some simple options in the library and leave an escape hatch for something like ssh where the amount of complexity is high and also all of this should be very much outside of the things we should have to care about.

Maybe at a later date we can have even built-in transports look like things you plug in so you can keep your own pointer to something if you need more than bog standard behaviour. But that's for when many other pressing problems have been solved.

carlosmn commented on src/transports/ssh_libssh.c:
```diff
@@ -0,0 +1,321 @@
+				assert(c->username);
+				assert(c->privatekey);
+
+				rc = libssh2_userauth_publickey_frommemory(
```
There are a series of `ssh_pki_import_privkey_` functions provided by libssh which let you specify an in-memory key in different formats. The one that would let us keep the same for both looks like it would be `ssh_pki_import_privkey_base64` as it oddly doesn't seem to allow you just pass the raw data, so we'd have to first convert it into base64.

carlosmn replied:
Or it looks like they might mean base64 as in the format you would find on-disk

tiennou reviewed (commented):
A few notes, sadly it's been a while so the context-switch is kinda cold on my side but that's what you get when you drop off the face of the earth.

Just to clarify, the libssh backend as it is should now handle "our" `keyboard-interactive` interface via a trampoline that just collects all challenges and converts them. It's just that actually setting up a test for that requires a harness/configuration which I haven't really looked at, so it's untested.

The thing that's (IIRC) completely missing from libssh is a way to support our custom hashing. Conversely, libssh seems to refer to GSSAPI mic authentication, which is absent from libssh2 (and gave me a good chuckle as a long-time Uplink player — I wasn't aware that voice authenticated SSH was a thing 😅).

I do agree that, if I was designing this from a blank slate, I'd go for a much more simple "try once with what the user gave us (pass, file/memory pkey, whatever, even a blanket this-asked-for-auth `void *` handle), or even and fail immediately with no retries", but API-wise that's not possible.

Anyway, I'll try to revisit that soonish, thanks for the review @carlosmn !

tiennou commented on src/transports/ssh_libssh.c:
```diff
@@ -0,0 +1,321 @@
+	unsigned char *hash_ptr = (unsigned char *)&hash;
+	size_t hash_len;
+
+	if (ssh_get_server_publickey(s->session, &pkey) == SSH_OK) {
```
Fixed. Note that I'm putting the base64-encoded raw data, because that's what libssh provides.

tiennou commented on src/transports/ssh_libssh2.c:
```diff
@@ -0,0 +1,437 @@
+int git_ssh_session_server_is_known(git_ssh_session *s, int *valid)
+{
+	GIT_UNUSED(s);
+	*valid = 0;
```
This ensures that libssh2's behavior doesn't change (default to "can't say it's valid", a.k.a `0`), but now libssh checks for known_hosts validity, while libssh2 doesn't (and it looks messy to handle appropriately). I'm not even sure if it makes sense to update the known_host file afterwards…

jkaiser-ioki commented:
Just to add my 2 cents. GitKraken (https://www.gitkraken.com/) is using libgit2, but due to the underlying usage of libssh2, it is currently not possible to use the client when talking to git remotes on ssh-hardened servers (e.g. only ed25519 host keys, or restricted allowed-cipher-sets, ...). Seeing this go through would probably improve things significantly on that front :-)

ethomson commented:
We have openssh support now — I really like libssh (and not so much libssh2) but it feels like openssh is the thing to use for "better openssh compatibility"
